modules/service/src/main/scala/ch/epfl/bluebrain/nexus/iam/service/auth/CredentialsStoreActor.scala

```diff
@@ -0,0 +1,118 @@
+package ch.epfl.bluebrain.nexus.iam.service.auth
+
+import java.security.PublicKey
+
+import akka.actor.{Actor, ActorLogging, ActorRef, ActorSystem, Props}
+import akka.cluster.singleton.{
+  ClusterSingletonManager,
+  ClusterSingletonManagerSettings,
+  ClusterSingletonProxy,
+  ClusterSingletonProxySettings
+}
+import akka.stream.ActorMaterializer
+import ch.epfl.bluebrain.nexus.commons.http.HttpClient.UntypedHttpClient
+import ch.epfl.bluebrain.nexus.commons.service.stream.SingletonStreamCoordinator.Stop
+import ch.epfl.bluebrain.nexus.iam.service.auth.CredentialsStoreActor.Protocol._
+import ch.epfl.bluebrain.nexus.iam.service.auth.TokenValidationFailure.KidOrIssuerNotFound
+import ch.epfl.bluebrain.nexus.iam.service.config.AppConfig.{OidcConfig, OidcProviderConfig}
+
+import scala.concurrent.{ExecutionContext, Future}
+import akka.actor.Status.Failure
+
+/**
+  * Actor implementation that maintains the keys of the OIDC providers defined on configuration.
+  * Those keys are going to be used to verify the JWT signature.
+  */
+class CredentialsStoreActor(providers: List[OidcProviderConfig])(implicit ucl: UntypedHttpClient[Future])
+    extends Actor
+    with ActorLogging {
+
+  private implicit val as: ActorSystem       = context.system
+  private implicit val ec: ExecutionContext  = context.dispatcher
+  private implicit val mt: ActorMaterializer = ActorMaterializer()
+  private var keys: Map[TokenId, PublicKey]  = Map()
+
+  override def preStart(): Unit =
+    updateKeys(providers)
+
+  def receive: Receive = {
+    case FetchKey(id) =>
+      val requester = sender()
+      val _ = keys.get(id) match {
+        case Some(key) =>
+          requester ! key
+        case None =>
+          requester ! Failure(KidOrIssuerNotFound)
+          log.warning("key for id '{}' not found on the provided keys", id)
+          updateKeys(providers)
+      }
+
+    case RefreshCredentials(provider) =>
+      updateKeys(List(provider))
+    // $COVERAGE-OFF$
+    case Stop =>
+      log.info("Received stop signal, stopping")
+      context.stop(self)
+    // $COVERAGE-ON$
+  }
+
+  private def updateKeys(provs: List[OidcProviderConfig]): Unit = {
+    val _ = provs
+      .foldLeft(Future(Map.empty[TokenId, PublicKey])) { (providersKey, provider) =>
+        val provKey = JwkClient(provider)
+          .map { key =>
+            log.info("key for the provider '{}' has been retrieved from {}", provider.issuer, provider.jwkCert)
+            key
+          }
+          .recoverWith {
+            log.warning("key for the provider '{}' has failed to be retrieved from {}",
+              provider.issuer,
+              provider.jwkCert)
+            Map()
+          }
+        (providersKey zip provKey).map {
+          case (acc, current) => acc ++ current
+        }
+      }
+      .map(keysMap => keys = keysMap)
```

Comment:
you could use `foreach` instead of `map` here, this way you wouldn't need `val _ =`